src/grfid/grfid.cpp

```diff
@@ -0,0 +1,289 @@
+/*
+ * This file is part of GRFCodec.
+ * GRFCodec is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * GRFCodec is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "../stdafx.h"
+
+#include "../3rdparty/md5/md5.h"
+#include "../3rdparty/fmt/format.h"
+
+#include <cstdio>
+#include <fcntl.h>
+
+#ifndef _WIN32
+#include <unistd.h>
+#include <sys/mman.h>
+#else
+#include <io.h>
+#endif
+
+#include "../safeguards.h"
+
+#define VERSION "openttd-vendored"
+
+struct MmapFile {
+	int fd = -1;
+	size_t file_length;
+	uint8_t *file_buffer;
+	uint8_t *buffer;
+
+	~MmapFile()
+	{
+		this->Close();
+	}
+
+	/**
+	 * Open a file for reading.
+	 * @param filename Filename of file to read.
+	 * @return true iff the file was successfully opened for reading.
+	 */
+	bool Open(const std::string &filename)
+	{
+		this->fd = open(filename.c_str(), O_RDONLY);
+		if (this->fd < 0) return false;
+
+		/* Determine length of file */
+		this->file_length = lseek(this->fd, 0, SEEK_END);
+		if (this->file_length == 0) return false;
+
+		lseek(this->fd, 0, SEEK_SET);
+
+#ifndef _WIN32
+		this->file_buffer = static_cast<uint8_t *>(mmap(NULL, this->file_length, PROT_READ, MAP_PRIVATE, fd, 0));
+		if (this->file_buffer == nullptr) return false;
+#else
+		this->file_buffer = new uint8_t[this->file_length];
+		if (this->file_buffer == nullptr) return false;
+		size_t file_read = read(this->fd, this->file_buffer, this->file_length);
```

Comment:
MSVC warnings are valid (the function expects a 32bit unsigned int) and `file_length` is 64bit.
But `lseek` return is a 32bit signed value, so an explicit cast should be fine.

Reply:
Who would ever have a 2GB file... but okay, unlikely for a NewGRF :D